bugsnag-plugin-android-anr/src/main/java/com/bugsnag/android/AnrPlugin.kt

```diff
@@ -42,31 +43,70 @@ internal class AnrPlugin : Plugin {
 
     override fun unload() = disableAnrReporting()
 
+    private fun getNotifyAnrDetectedIndex(stackTrace: Array<StackTraceElement>): Int {
+        for (i in 0..stackTrace.size) {
+            val frame = stackTrace[i]
+            if (frame.methodName == "notifyAnrDetected" && frame.className == "com.bugsnag.android.AnrPlugin") {
+                return i
+            }
+        }
+        return -1
+    }
+
+    private fun isNativeANR(stackTrace: Array<StackTraceElement>): Boolean {
+        // True if the frame leading to "notifyAnrDetected" is native.
+        // This is not perfect, because some JVM calls are technically native (like sleep).
+
+        val notifyIndex = getNotifyAnrDetectedIndex(stackTrace)
+        if (notifyIndex < 0) {
+            return false
+        }
+
+        return stackTrace[notifyIndex + 1].isNativeMethod
+    }
+
+    private fun mergeTraces(jvmTrace: Array<StackTraceElement>, nativeTrace: List<Stackframe>):
+        Array<StackTraceElement> {
+            val notifyIndex = getNotifyAnrDetectedIndex(jvmTrace)
+
+            val stackTrace = LinkedList<StackTraceElement>()
+            for (frame in nativeTrace) {
+                stackTrace.add(
+                    StackTraceElement(
+                        "",
+                        frame.method,
+                        frame.file,
+                        frame.lineNumber?.toInt() ?: 0
+                    )
+                )
+            }
+            for (i in notifyIndex + 1..jvmTrace.size - 1) {
+                stackTrace.add(jvmTrace[i])
+            }
+            return stackTrace.toTypedArray()
+        }
+
     /**
      * Notifies bugsnag that an ANR has occurred, by generating an Error report and populating it
      * with details of the ANR. Intended for internal use only.
      */
     @Suppress("UNUSED_PARAMETER")
     private fun notifyAnrDetected(info: Long, userContext: Long) {
-        val thread = Looper.getMainLooper().thread
-
         // generate a full report as soon as possible, then wait for extra process error info
-        val exc = RuntimeException()
```

Comment:
this shouldn't be moved unless there's a good reason to do so as it affects the accuracy of the stacktrace?

Reply:
The real stack trace is pulled from `Looper.getMainLooper().thread.stackTrace` at the beginning of the method. The `RuntimeException` is just a holder for this trace (its `stackTrace` var gets replaced with the trace grabbed from the main thread) so that the contrived exception can be converted to the internal trace format via the standard path.

Reply:
Ah, got it - this is fine then! :)